odoo/tools/profiler.py

```diff
@@ -1,197 +1,401 @@
+        self.init_frame = get_current_frame(self.init_thread)
+        if self.init_frame.f_code.co_name == 'enter_context':
+            self.init_frame = self.init_frame.f_back  # profiler used in a ExitStack case
+        self.init_stack_trace = _get_stack_trace(self.init_frame)
+        if self.description is None:
+            frame = self.init_frame
+            code = frame.f_code
+            self.description = f"{frame.f_code.co_name} ({code.co_filename}:{frame.f_lineno})"
+        if self.disable_gc and gc.isenabled():
+            gc.disable()
+        self.start_time = time.time()
+        for collector in self.collectors:
+            collector.start()
+        return self
+
+    def __exit__(self, *args):
+        try:
+            for collector in self.collectors:
+                collector.stop()
+            self.duration = time.time() - self.start_time
+            self._add_file_lines(self.init_stack_trace)
+            if self.path:
+                save_time = datetime.datetime.now().strftime("%Y%m%d-%H%M%S")
+                dirname = os.path.dirname(self.path)
+                if not os.path.exists(dirname):
+                    os.makedirs(dirname)
+                for collector in self.collectors:
+                    formatted_path = self.path.format(len=len(collector.entries), time=save_time)
+                    row = collector.name
+                    path = f'{formatted_path}_{self.description}_{row}.json'
+                    _logger.warning('saving to %s', path)
+                    with open(path, 'w') as f:
+                        json.dump({
+                            "name": self.description,
+                            "session": self.profile_session,
+                            "create_date": save_time,
+                            "init_stack_trace": _format_stack(self.init_stack_trace),
+                            "duration": self.duration,
+                            'result': collector.entries,
+                        }, f, indent=4)
+
+            if self.db:
+                # pylint: disable=import-outside-toplevel
+                from odoo.sql_db import db_connect  # only import from odoo if/when needed.
+                with db_connect(self.db).cursor() as cr:
+                    values = {
+                        "name": self.description,
+                        "session": self.profile_session,
+                        "create_date": datetime.datetime.now(),
+                        "init_stack_trace": json.dumps(_format_stack(self.init_stack_trace)),
+                        "duration": self.duration,
+                    }
+                    for collector in self.collectors:
+                        if collector.entries:
+                            values[collector.name] = json.dumps(collector.entries)
+                    cols = ",".join(values)
+                    query = f"INSERT INTO ir_profile({cols}) VALUES %s RETURNING id"
```

Comment:
Use psycopg2.sql helpers? I'm surprised the linter doesn't reject this.

Reply:
Second parameter is given to the `execute` function so the linter assumes you know what you are doing.